### PR TITLE
feat(portals): React router child routes

### PR DESCRIPTION
**Pull request by snaerseljan**
# React router updates

## What

Let me start to apologise for the size of the PR 😇 . I was working on these features while another important RR branch was not merged. Big part of the PR is because of changes in each module to use `element` prop instead of `render` fn.

This PR tackles the final big update of React Router. The following was implemented.
- All portal modules now use React Router properties along with our own. See ´portals/core/Readme.md´.
- ADS now uses loader function to preload necessary data before rendering route. This is the proof of concept with the loader function.
- Both portals now have a page loader at the top of the page for indicating if router is loading new route. This is mainly to show some loading indication when the loader function is executed.
- Router now shows loading screen before everything is loaded.
- Child route support. Module routes can now have child routes which render inside of the route.
- And some refactor of course.

## Why

We want to fully support RR new features 

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/portals/admin/air-discount-scheme/src/screens/Overview/Overview.tsx [hunk 1]
```diff
@@ -23,48 +23,37 @@ import {
...
   }
+
   const [
     confirmInvoice,
     { loading: confirmInvoiceLoading },
   ] = useConfirmInvoiceMutation()
 
   const { data, loading: queryLoading, refetch } = useFlightLegsQuery({
     ssr: false,
-    fetchPolicy: 'network-only',
```
Comment (snaerseljan) [inline libs/portals/admin/air-discount-scheme/src/screens/Overview/Overview.tsx:67]: Removed in favour of network call in loader.

File: libs/portals/core/README.md [hunk 2]
```diff
@@ -50,27 +49,58 @@ export interface PortalModuleProps {
...
 
 ```typescript
-export type PortalRoute = {
+import { RouteObject } from 'react-router-dom'
+
+export type PortalRoute = RouteObject & {
   name: string
   path: string | string[]
-  render?: (props: PortalModuleProps) => PortalModuleRenderValue
+  element?: React.ReactNode
```
Comment (eirikurn) [inline libs/portals/core/README.md:57]: Remove?

File: libs/portals/core/README.md [hunk 3]
```diff
@@ -50,27 +49,58 @@ export interface PortalModuleProps {
...
+  /**
+   * Route actions are the "writes" to route loader "reads". 
+   * They provide a way for apps to perform data mutations with simple HTML and HTTP semantics 
+   * while React Router abstracts away the complexity of asynchronous UI and revalidation.
+   */
+  action?: RouteObject['action']
+  /**
+   * The children prop is rendered when the path matches. Remember to use <Outlet /> in the parent compoennt to render the children.
+   */
+  children?: RouteObject['children']
```
Comment (eirikurn) [inline libs/portals/core/README.md:84]: Should be PortalRoute instead?

File: libs/portals/core/README.md [hunk 4]
```diff
@@ -50,27 +49,58 @@ export interface PortalModuleProps {
...
+const ProtectedScreen = lazy(() => import('./screens/ProtectedScreen/ProtectedScreen'))
+
 routes: (userInfo) => {
   const applicationRoutes = [
     {
       name: 'Applications',
       path: ServicePortalPath.ApplicationRoot,
-      render: () =>
-        lazy(() => import('./screens/ApplicationList/ApplicationList')),
+      element: <ApplicationList userInfo={userInfo} />,
```
Comment (eirikurn) [inline libs/portals/core/README.md:103]: Should stop promoting this pattern and instead encourage using a hook?

File: libs/portals/core/src/components/PageLoader.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+  const colorGradiant = `linear-gradient(90deg,
+    ${theme.color.blue400} 0%,
+    ${theme.color.blue600} 25%,
+    ${theme.color.purple400} 50%,
+    ${theme.color.roseTinted400} 75%,
+    ${theme.color.red400} 100%
+  )`
+
+  return <LoadingBar color={colorGradiant} ref={ref} />
```
Comment (eirikurn) [inline libs/portals/core/src/components/PageLoader.tsx:26]: ~Would be nice to move the visual part of this component to island-ui. Could even have a prop to start/stop to make it easier to integrate into our projects.

File: libs/portals/core/src/screens/ModuleErrorScreen.tsx [hunk 6]
```diff
@@ -3,48 +3,26 @@ import { Box, Text } from '@island.is/island-ui/core'
...
-    if (this.state.hasError) return <ModuleErrorScreen name={this.props.name} />
-    return this.props.children
-  }
-}
-
 export const ModuleErrorScreen = ({ name }: ModuleErrorScreenProps) => {
   const { formatMessage } = useLocale()
+  const error = useRouteError()
+
+  console.error(error)
```
Comment (eirikurn) [inline libs/portals/core/src/screens/ModuleErrorScreen.tsx:16]: Remove? Or at least move into effect to not log during render.

File: libs/portals/core/src/screens/ModuleRoute.tsx [hunk 7]
```diff
@@ -1,54 +1,45 @@
...
-      <Box paddingY={1}>
-        <ModuleErrorScreen name={route.name} />
-      </Box>
+      <Suspense fallback={null}>
+        <Box paddingY={1}>{module}</Box>
+      </Suspense>
     )
-  },
-)
+  }
```
Comment (eirikurn) [inline libs/portals/core/src/screens/ModuleRoute.tsx:38]: Checking `route.element` twice? Do we even need this check? Can probably get rid of the ModuleErrorScreen in this case.

**Review by eirikurn — COMMENTED** [review]
Awesome job!

**Review by eirikurn — APPROVED** [review]

File: apps/portals/admin/src/components/Layout/Layout.tsx [hunk 8]
```diff
@@ -40,35 +40,39 @@ const getGridColumnSize = (layout: PortalModule['layout']) => {
...
-}) => {
-  const hasNoneLayout = layout === 'none'
+type LayoutModuleContainerProps = {
+  layout: PortalModule['layout']
+}
 
-  if (hasNoneLayout) {
-    return <Box {...boxProps}>{children}</Box>
-  }
+const LayoutModuleContainer: FC<LayoutModuleContainerProps> = React.memo(
```
[the remark this answers is not in the document]
Comment (snaerseljan) [reply apps/portals/admin/src/components/Layout/Layout.tsx:47]: I noticed there where 3 re-renders. This takes it down to one

File: libs/portals/admin/air-discount-scheme/src/lib/loaders.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        from: -1,
+        to: 1000,
+      },
+      isExplicit: false,
+    },
+  }
+}
+
+export const overviewLoader: WrappedLoaderFn = ({ client }) => async () => {
+  const { input } = prepareFlightLegsQuery()
```
[the remark this answers is not in the document]
Comment (snaerseljan) [reply libs/portals/admin/air-discount-scheme/src/lib/loaders.ts:45]: Discussed offline. Decided to do a follow up PR

[the remark this answers is not in the document]
Comment (snaerseljan) [reply libs/portals/admin/air-discount-scheme/src/lib/loaders.ts:45]: But did update to useDataLoader

**Review by egillben — APPROVED** [review]
Like the lazy loading :) code owned files =  LGTM

File: libs/portals/core/src/components/PageLoader.tsx (hunk 5 above)
Comment (snaerseljan) [reply libs/portals/core/src/components/PageLoader.tsx:26]: Yeah I was thinking that as well but then I got lazy 😇 . I will do it now

**Review by thordurhhh — APPROVED** [review]
Service portal looking good, tested locally and everything I tried seemed to go smoothly 🙌

**Review by rafnarnason — APPROVED** [review]
lgtm

**Review by kksteini — APPROVED** [review]

**Review by Toti91 — APPROVED** [review]
UI files look good

**Pull request merged**